src/big-int/bigint.hh

```diff
@@ -236,109 +238,51 @@ public:
   BigInt &negate()			{ if(!is_zero()) positive = !positive; return *this; }
   BigInt operator-() const		{ return BigInt (*this).negate(); }
 
-  BigInt &operator+= (llong_t) _fast;
-  BigInt &operator-= (llong_t) _fast;
-  BigInt &operator*= (llong_t) _fast;
-  BigInt &operator/= (llong_t) _fast;
-  BigInt &operator%= (llong_t) _fast;
-
-  BigInt &operator= (unsigned long x) { return (*this)=(ullong_t)x; }
-  BigInt &operator+= (unsigned long x) { return (*this)+=(ullong_t)x; }
-  BigInt &operator-= (unsigned long x) { return (*this)-=(ullong_t)x; }
-  BigInt &operator*= (unsigned long x) { return (*this)*=(ullong_t)x; }
-  BigInt &operator/= (unsigned long x) { return (*this)/=(ullong_t)x; }
-  BigInt &operator%= (unsigned long x) { return (*this)%=(ullong_t)x; }
-
-  BigInt &operator+= (ullong_t) _fast;
-  BigInt &operator-= (ullong_t) _fast;
-  BigInt &operator*= (ullong_t) _fast;
-  BigInt &operator/= (ullong_t) _fast;
-  BigInt &operator%= (ullong_t) _fast;
-
-  BigInt &operator+= (BigInt const &) _fast;
-  BigInt &operator-= (BigInt const &) _fast;
-  BigInt &operator*= (BigInt const &) _fast;
-  BigInt &operator/= (BigInt const &) _fasta;
-  BigInt &operator%= (BigInt const &) _fasta;
+#define IN_PLACE_OPERATOR(TYPE) \
+  BigInt &operator+= (TYPE) _fast; \
+  BigInt &operator-= (TYPE) _fast; \
+  BigInt &operator*= (TYPE) _fast; \
+  BigInt &operator/= (TYPE) _fast; \
+  BigInt &operator%= (TYPE) _fast;
+
+  IN_PLACE_OPERATOR(const BigInt &)
+  IN_PLACE_OPERATOR(llong_t)
+  IN_PLACE_OPERATOR(ullong_t)
+#undef IN_PLACE_OPERATOR
+
+#define OVERLOAD_IN_PLACE_OPERATOR(FROM, TO) \
+  BigInt &operator+=(FROM x) { return operator+=(static_cast<TO>(x)); } \
+  BigInt &operator-=(FROM x) { return operator-=(static_cast<TO>(x)); } \
+  BigInt &operator*=(FROM x) { return operator*=(static_cast<TO>(x)); } \
+  BigInt &operator/=(FROM x) { return operator/=(static_cast<TO>(x)); } \
+  BigInt &operator%=(FROM x) { return operator%=(static_cast<TO>(x)); }
+
+  OVERLOAD_IN_PLACE_OPERATOR(unsigned long, ullong_t)
+  OVERLOAD_IN_PLACE_OPERATOR(int, llong_t)
+  OVERLOAD_IN_PLACE_OPERATOR(unsigned, ullong_t)
+#undef OVERLOAD_IN_PLACE_OPERATOR
```

Comment:
Could these macros be implemented as free template operator overloads with `typename std::enable_if<std::is_integral<T>::value>::type`?